numba/tests/test_struct_ref.py

```diff
@@ -0,0 +1,273 @@
+        self.assertNotIsInstance(first_got, MyStruct)
+        self.assertPreciseEqual(first_expected, get_values(first_got))
+
+        second_expected = first_expected + (ctr * ctr)
+        second_got = compute_fields(first_got)
+        self.assertPreciseEqual(second_expected, second_got)
+
+    def test_MySimplerStructType_wrapper_has_no_attrs(self):
+        vs = np.arange(10, dtype=np.intp)
+        ctr = 13
+        wrapper = ctor_by_intrinsic(vs, ctr)
+        self.assertIsInstance(wrapper, structref.StructRefProxy)
+        with self.assertRaisesRegex(AttributeError, 'values'):
+            wrapper.values
+        with self.assertRaisesRegex(AttributeError, 'counter'):
+            wrapper.counter
+
+    def test_MyStructType(self):
+        vs = np.arange(10, dtype=np.float64)
+        ctr = 11
+
+        first_expected_arr = vs.copy()
+        first_got = ctor_by_class(vs, ctr)
+        self.assertIsInstance(first_got, MyStruct)
+        self.assertPreciseEqual(first_expected_arr, first_got.values)
+
+        second_expected = first_expected_arr + ctr
+        second_got = compute_fields(first_got)
+        self.assertPreciseEqual(second_expected, second_got)
+        self.assertEqual(first_got.counter, ctr)
+
+    def test_MyStructType_mixed_types(self):
+        # structref constructor is generic
+        @njit
+        def mixed_type(x, y, m, n):
+            return MyStruct(x, y), MyStruct(m, n)
+
+        a, b = mixed_type(1, 2.3, 3.4j, (4,))
+        self.assertEqual(a.values, 1)
+        self.assertEqual(a.counter, 2.3)
+        self.assertEqual(b.values, 3.4j)
+        self.assertEqual(b.counter, (4,))
+
+    def test_MyStructType_in_dict(self):
+        td = typed.Dict()
+        td['a'] = MyStruct(1, 2.3)
+        self.assertEqual(td['a'].values, 1)
+        self.assertEqual(td['a'].counter, 2.3)
+        # overwrite
+        td['a'] = MyStruct(2, 3.3)
+        self.assertEqual(td['a'].values, 2)
+        self.assertEqual(td['a'].counter, 3.3)
+        # mutate
+        td['a'].values += 10
+        self.assertEqual(td['a'].values, 12)    # changed
+        self.assertEqual(td['a'].counter, 3.3)  # unchanged
+
+    def test_MyStructType_in_dict_mixed_type_error(self):
+        self.disable_leak_check()
+
+        td = typed.Dict()
+        td['a'] = MyStruct(1, 2.3)
+        self.assertEqual(td['a'].values, 1)
+        self.assertEqual(td['a'].counter, 2.3)
+
+        # ERROR: store different types
+        with self.assertRaisesRegex(errors.TypingError,
+                                    r"Cannot cast numba.MyStructType"):
+            # because first field is not a float;
+            # the second field is now an integer.
+            td['b'] = MyStruct(2.3, 1)
+
+
+@overload_method(MyStructType, "testme")
+def _(self, arg):
+    def impl(self, arg):
+        return self.values * arg + self.counter
+    return impl
+
+
+class TestStructRefExtending(MemoryLeakMixin, TestCase):
+    def test_overload_method(self):
```

Comment:
RE overloads, is attribute overloading likely to ever be a thing that makes sense in the `structref` context?

Reply:
I guess to emulate `@property` without an actual field backing it. I can add a test quickly